src/tree_cmd.cpp

```diff
@@ -178,6 +179,108 @@ static void PlaceTree(TileIndex tile, uint32_t r)
+	float phase;
+	int frequency;
+};
+
+/**
+ * Creates a star-shaped polygon originating from (0, 0) as defined by the given harmonics.
+ * @param radius The maximum radius of the polygon. May be smaller, but will not be larger.
+ * @param harmonics Harmonics data for the polygon.
+ * @param shape Shape to fill with points.
+ */
+static void CreateStarShapedPolygon(const int radius, std::span<const BlobHarmonic> harmonics, std::span<Point> shape)
+{
+	float theta = 0;
+	float step = (M_PI * 2) / std::size(shape);
+
+	/* Divide a circle into a number of equally spaced divisions. */
+	for (Point &vertex : shape) {
+
+		/* Add up the values of each harmonic at this segment.*/
+		float deviation = std::accumulate(std::begin(harmonics), std::end(harmonics), 0, [theta](float d, const BlobHarmonic &harmonic) {
+			return d + sin((theta + harmonic.phase) * harmonic.frequency) * harmonic.amplitude;
+		});
+
+		/* Smooth out changes. */
+		float adjusted_radius = (radius / 2.0) + (deviation / 2);
+
+		/* Add to the final polygon. */
+		vertex.x = cos(theta) * adjusted_radius;
+		vertex.y = sin(theta) * adjusted_radius;
+
+		/* Proceed to the next segment. */
+		theta += step;
+	}
+}
+
+/**
+ * Creates a random star-shaped polygon originating from (0, 0).
+ * @param radius The maximum radius of the blob. May be smaller, but will not be larger.
+ * @param[out] shape Shape to fill with polygon points.
+ */
+static void CreateRandomStarShapedPolygon(int radius, std::span<Point> shape)
+{
+	/* Valid values for the phase of blob harmonics are between 0 and Tau. we can get a value in the correct range
+	 * from Random() by dividing the maximum possible value by the desired maximum, and then dividing the random
+	 * value by the result. */
+	static constexpr float PHASE_DIVISOR = INT32_MAX / (M_PI * 2);
+
+	/* These values are ones found in testing that result in suitable-looking polygons that did not self-intersect
+	 * and fit within a square of radius * radius dimensions. */
+	std::initializer_list<BlobHarmonic> harmonics = {
+		{radius / 2, Random() / PHASE_DIVISOR, 1},
+		{radius / 4, Random() / PHASE_DIVISOR, 2},
+		{radius / 8, Random() / PHASE_DIVISOR, 3},
+		{radius / 16, Random() / PHASE_DIVISOR, 4},
+	};
+
+	CreateStarShapedPolygon(radius, harmonics, shape);
+}
+
+/**
+ * Returns true if the given coordinates lie within a triangle.
+ * @param x X coordinate relative to centre of shape.
+ * @param y Y coordinate relative to centre of shape.
+ * @param v1 First vertex of triangle.
+ * @param v2 Second vertex of triangle.
+ * @param v3 Third vertex of triangle.
+ * @returns true if the given coordinates lie within a triangle.
+ */
+static bool IsPointInTriangle(const int x, const int y, const Point &v1, const Point &v2, const Point &v3)
+{
+	const int s = ((v1.x - v3.x) * (y - v3.y)) - ((v1.y - v3.y) * (x - v3.x));
+	const int t = ((v2.x - v1.x) * (y - v1.y)) - ((v2.y - v1.y) * (x - v1.x));
+
+	if (s < 0 != t < 0 && s != 0 && t != 0) return false;
```

Comment:
```suggestion
	if ((s < 0) != (t < 0) && s != 0 && t != 0) return false;
```

just a little clearer